File: OCDemo/TableViewDemo/TableViewDemo/ViewController.m
//
//  ViewController.m
//  TableViewDemo
//
//  Created by admin on 2022/9/20.
//

#import "ViewController.h"
#import "XBCarGroup.h"
#import "XBCar.h"

@interface ViewController ()

@property(nonatomic, weak) UITableView *tableView1;

@property (nonatomic, strong) NSArray *carGroups;

@end

@implementation ViewController

- (NSArray *)carGroups{
    if(!_carGroups){
//        从plist文件加载数据
        NSArray *dictArray = [NSArray arrayWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"Cars" ofType:@"plist"]];

        NSMutableArray *temp = [NSMutableArray array];
        for(NSDictionary *carGroupDict in dictArray) {
            XBCarGroup *group = [XBCarGroup carGroupWithDict: carGroupDict];
            [temp addObject:group];
        }
        _carGroups = temp;
        
//        XBCarGroup *group0 = [[XBCarGroup alloc] init];
//        group0.header = @"德系";
//        group0.footer = @"德系质量好ooo";
//        group0.cars = @[
//            [XBCar carWithName:@"Benz" icon:@"p01"],
//            [XBCar carWithName:@"BMW" icon:@"p02"],
//        ];
//        XBCarGroup *group1 = [[XBCarGroup alloc] init];
//        group1.header = @"日系";
//        group1.footer = @"日系价格便宜000";
//        group1.cars = @[
//            [XBCar carWithName:@"丰田" icon:@"p03"],
//            [XBCar carWithName:@"本田" icon:@"p04"],
//            [XBCar carWithName:@"马自达" icon:@"p05"],
//        ];
//        _carGroups = @[group0, group1];
    }
    return _carGroups;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    UITableView *tableView = [[UITableView alloc] initWithFrame:self.view.frame style: UITableViewStyleGrouped];
    self.tableView1 = tableView;
    self.tableView1.delegate = self;
    self.tableView1.dataSource = self;
//    常见属性设置
    self.tableView1.rowHeight = 80;
    self.tableView1.sectionHeaderHeight = 80;
    self.tableView1.separatorColor = [UIColor clearColor];
    self.tableView1.tableHeaderView = [[UISwitch alloc] init];
    self.tableView1.tableFooterView = [[UISwitch alloc] init];
    
    
    [self.view addSubview: self.tableView1];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return self.carGroups.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    XBCarGroup *group = self.carGroups[section];
    return group.cars.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    
//    优化
    NSString *ID = @"Car";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:ID];
    if(cell == nil){
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:ID];
        cell.accessoryView = [[UISwitch alloc] init];
        UIView *bg = [[UIView alloc] init];
        bg.backgroundColor = [UIColor yellowColor];
        cell.selectedBackgroundView = bg;
    }
//    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
//    常见属性设置
   
    
    XBCarGroup *group = self.carGroups[indexPath.section];
    XBCar *car = group.cars[indexPath.row];
    
    cell.textLabel.text = car.name;
    cell.imageView.image = [UIImage imageNamed:car.icon];
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section{
    XBCarGroup *group = self.carGroups[section];
    return group.header;
}

- (NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section{
    XBCarGroup *group = self.carGroups[section];
    return group.footer;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    NSLog(@"choose:%ld",indexPath.row);
}

@end

File: daysDemo/day05_PullToRefresh/PullToRefresh/PullToRefresh/ViewController.swift
//
//  ViewController.swift
//  PullToRefresh
//
//  Created by admin on 2022/7/30.
//

import UIKit

// 注意继承和协议，table使用所必须得
class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    
    var dataSource = Array<Date>()
    let refresh = UIRefreshControl()
    var table: UITableView!

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        table = UITableView(frame: self.view.bounds)
        table.frame.origin.y = 44
        self.view.addSubview(table)
        table.delegate = self // table使用必需
        table.dataSource = self // table list数据源
        addNewElementToArray()
        refresh.attributedTitle = NSAttributedString(string: "正在下拉刷新。。。") // 刷新时头部显示的文案
        
        refresh.addTarget(self, action: #selector(pullTheRefresh), for: .valueChanged)
        
        self.table.refreshControl = refresh
    }
    
    @objc func pullTheRefresh() {   // 下拉刷新后执行的动作
        addNewElementToArray()
        refresh.endRefreshing()
        table.reloadData()
    }
    
    func addNewElementToArray() {
        dataSource.insert(NSDate() as Date, at: 0)
    }

    func numberOfSections(in tableView: UITableView) -> Int {  // table子元素里面section的个数， Table中包含一个section，这一个section中包含dataSource.count个row
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {  // table子元素里面每个section包含row的个数
        return dataSource.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {  // row中cell的显示样式
        let cell = UITableViewCell(style: .default, reuseIdentifier: "reuseCellForPullToRefresh")
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy年MM月dd日/HH时mm分ss秒"
        let dateStr = dateFormatter.string(from: dataSource[indexPath.row])
        cell.textLabel?.text = dateStr
        return cell
    }
}

